unified_planning/io/pddl_writer.py

```diff
@@ -146,31 +146,47 @@ def _write_domain(self, out: IO[str]):
+            if problem_kind.has_equality(): # type: ignore
                 out.write(' :equality')
-            if (self.problem.kind().has_continuous_numbers() or # type: ignore
-                self.problem.kind().has_discrete_numbers()): # type: ignore
+            if (problem_kind.has_continuous_numbers() or # type: ignore
+                problem_kind.has_discrete_numbers()): # type: ignore
                 out.write(' :numeric-fluents')
-            if self.problem.kind().has_conditional_effects(): # type: ignore
+            if problem_kind.has_conditional_effects(): # type: ignore
                 out.write(' :conditional-effects')
-            if self.problem.kind().has_existential_conditions(): # type: ignore
+            if problem_kind.has_existential_conditions(): # type: ignore
                 out.write(' :existential-preconditions')
-            if self.problem.kind().has_universal_conditions(): # type: ignore
+            if problem_kind.has_universal_conditions(): # type: ignore
                 out.write(' :universal-preconditions')
-            if (self.problem.kind().has_continuous_time() or # type: ignore
-                self.problem.kind().has_discrete_time()): # type: ignore
+            if (problem_kind.has_continuous_time() or # type: ignore
+                problem_kind.has_discrete_time()): # type: ignore
                 out.write(' :durative-actions')
-            if self.problem.kind().has_duration_inequalities(): # type: ignore
+            if problem_kind.has_duration_inequalities(): # type: ignore
                 out.write(' :duration-inequalities')
             out.write(')\n')
 
-        out.write(f' (:types {" ".join([t.name() for t in self.problem.user_types()])})\n' if len(self.problem.user_types()) > 0 else '') # type: ignore
+        if problem_kind.has_hierarchical_typing(): # type: ignore
+            self.object_freshname = 'object'
+            while self.problem.has_type(self.object_freshname):
+                self.object_freshname = self.object_freshname + '_'
+            user_types_hierarchy = self.problem.user_types_hierarchy()
+            out.write(f' (:types\n')
+            stack: List['unified_planning.model.Type'] = user_types_hierarchy[None] if None in user_types_hierarchy else []
+            out.write(f'    {" ".join(t.name() if t.name() != "object" else self.object_freshname for t in stack)} - object\n')  # type: ignore
+            while stack:
+                current_type = stack.pop()
+                direct_sons: List['unified_planning.model.Type'] = user_types_hierarchy[current_type]
+                if direct_sons:
+                    stack.extend(direct_sons)
+                    out.write(f'    {" ".join([t.name() if t.name() != "object" else self.object_freshname for t in direct_sons])} - {current_type.name() if current_type.name() != "object" else self.object_freshname}\n')  # type: ignore
+            out.write(' )\n')
+        else:
```

Comment:
Is this else needed? if there is no type hierarchy the code above should work already. Moreover, you are not handling the case of no-hierarchy with a type called `object`